File: ch3/rsa-encrypt.py
#
# Encrypt with public key
#
# From: "The Chains that bind us" by Phillip G Bradford
#  https://github.com/wonder-phil/BlockchainTechNotes
#      by P. G. Bradford
#
#
# Do the following in order
#
# 1 Computer> python3 key-gen.py
# 2 Computer> python3 rsa-encrypt.py
# 3 Computer> python3 rsa-decrypt.py
#
import rsa

public_key_file = open("public-key.pem","r")
public_key_data = public_key_file.read().encode('utf-8')
public_key_file.close()

print(public_key_data)

public_key = rsa.PublicKey.load_pkcs1(bytes(public_key_data),'PEM')
data = "Greetings - Today is a great day!".encode("utf-8")

encrypted = rsa.encrypt(data,public_key)

file_out = open("encrypted_data.bin", "wb")

File: cpp/Ch3/rsa_encrypt.cpp
#include <iostream>
#include <fstream>
#include <openssl/pem.h>
#include <openssl/rsa.h>
#include <openssl/err.h>

RSA* loadPublicKey(const std::string& filename) {
    FILE* file = fopen(filename.c_str(), "rb");
    if (!file) {
        std::cerr << "Unable to open public key file: " << filename << std::endl;
        return nullptr;
    }

    RSA* rsa = PEM_read_RSA_PUBKEY(file, nullptr, nullptr, nullptr);
    fclose(file);

    if (!rsa) {
        std::cerr << "Failed to load public key from: " << filename << std::endl;
    }

    return rsa;
}

std::vector<unsigned char> encrypt(RSA* rsa, const std::string& data) {
    if (!rsa) {
        std::cerr << "RSA context is null" << std::endl;
        return {};
    }

    int rsaSize = RSA_size(rsa);
    std::vector<unsigned char> encrypted(rsaSize);

    int result = RSA_public_encrypt(data.size(), 
                                    reinterpret_cast<const unsigned char*>(data.c_str()),
                                    encrypted.data(), rsa, RSA_PKCS1_PADDING);

    if (result == -1) {
        unsigned long err = ERR_get_error();
        char err_msg[1024] = {};
        ERR_error_string_n(err, err_msg, sizeof(err_msg));
        std::cerr << "Error encrypting message: " << err_msg << std::endl;
        return {};
    }

    encrypted.resize(result);
    return encrypted;
}

int main() {
    const std::string publicKeyFile = "public-key.pem";
    RSA* rsa = loadPublicKey(publicKeyFile);
    if (!rsa) {
        return 1;
    }

    const std::string data = "Greetings - Today is a great day!";
    std::vector<unsigned char> encrypted = encrypt(rsa, data);

    std::ofstream fileOut("encrypted_data.bin", std::ios::binary);
    if (!fileOut) {
        std::cerr << "Unable to open file for writing encrypted data." << std::endl;
        RSA_free(rsa);
        return 1;
    }

    fileOut.write(reinterpret_cast<const char*>(encrypted.data()), encrypted.size());
    fileOut.close();

    RSA_free(rsa);
    return 0;
}
